### DISCUSSION TITLE
Excluding None (optional fields) from json response bodies

### DISCUSSION THREAD

**Discussion by bunny-therapist** [litestar-org/litestar#3072] [Q&A] [answered]
Many of our pydantic models have optional fields (`field: Optional[T] = None`). Those fields are not listed as required in the equivalent json schema. When I use those models as response bodies with litestar, those fields get included with a null value in the serialized response. How can I exclude those fields completely from the body?

When doing this with pydantic natively, I can just pass `exclude_none=True` to `model_dump`. When building an app with FastAPI I can set `responce_model_exclude_none=True` for the endpoint. With litestar, I cannot figure out any way to do this other than to just convert the model to a dict myself and return that. However, that seems to go against the intention of the framework...

Is there any way to exclude optional fields from the serialized json response?

**Comment by bunny-therapist** [comment]
On a related note, I have a similar concern regarding excluding default values (`exclude_defaults=True`).

I simply cannot see how to customize the serialization of the pydantic models (other than develop some kind of plugin I suppose, but that seems very complicated).

**Comment by bunny-therapist** [comment]
Is `type_encoders` the recommended solution?

**Reply by provinzkraut** [reply, marked as answer]
That's the best way to do this for Pydantic, yes.
